### PR TITLE
#748 add oportunity for getting classbook in panel

### LINKED ISSUE #748: Add dashboard (Panel)
In API we have endpoints for getting classbook data from database:

image

There has to be an opportunity just like in API

### PR DESCRIPTION
**Pull request by DmitryTkachenko**

### CODE REVIEW

File: CharlieBackend.Panel/Controllers/ClassbookController.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+﻿using CharlieBackend.Panel.Services.Interfaces;
+using CharlieBackend.Panel.Models.Classbook;
+using Microsoft.AspNetCore.Authorization;
+using CharlieBackend.Core.DTO.Dashboard;
+using CharlieBackend.Panel.Services.Interfaces;
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Panel/Controllers/ClassbookController.cs:3]: please remove

File: CharlieBackend.Panel/Controllers/ClassbookController.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+using Microsoft.AspNetCore.Authorization;
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Panel/Controllers/ClassbookController.cs:4]: please remove

File: CharlieBackend.Panel/Controllers/ClassbookController.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        public async Task<IActionResult> Index()
+        {
+            var param = new StudentsRequestDto<ClassbookResultType>();
+            ClassbookViewModel data = await _classbookService.GetClassbookAsync(param);
+
+            return View(data);
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> GetClassbook(StudentsRequestDto<ClassbookResultType> scheduledEventFilter)
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Panel/Controllers/ClassbookController.cs]: please add async to naming

File: CharlieBackend.Panel/Services/ClassbookService.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+        /// <summary>
+        /// Applies default classbook filter.
+        /// </summary>
+        /// <param name="classbookFilter">Object with a set of 
+        /// optional filtration parameters to add default values to.</param>
+        private void ApplyDefaultClassbookFilter(
+            StudentsRequestDto<ClassbookResultType> classbookFilter)
+        {
+            if (classbookFilter.StartDate==default(DateTime))
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Panel/Services/ClassbookService.cs]: please add space

File: CharlieBackend.Panel/Services/ClassbookService.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            {
+                classbookFilter.StartDate = new DateTime(DateTime.Now.Year, 1, 1);
+            }
+            if (classbookFilter.FinishDate==default(DateTime))
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Panel/Services/ClassbookService.cs]: same here and in next

File: CharlieBackend.Panel/Services/ClassbookService.cs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+        private async Task<IList<ClassbookStudentsGroupViewModel>> GetStudentsGroupViewModelsAsync()
+        {
+            var studentGroupDtos = 
+                await _apiUtil.GetAsync<IList<StudentGroupDto>>(_getStudentGroupsEndpoint);
+
+            return _mapper.Map<IList<ClassbookStudentsGroupViewModel>>(studentGroupDtos);
+        }
+        
+        private async Task<FilteredClassbookViewModel> GetFilteredRegisterViewModelAsynk(
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Panel/Services/ClassbookService.cs]: please rename asynk => async

File: CharlieBackend.Panel/Services/ClassbookService.cs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            StudentsRequestDto<ClassbookResultType> groupsRegisterFilter)
+        {
+            var result = await _apiUtil.PostAsync< StudentsClassbookResultDto, StudentsRequestDto <ClassbookResultType>>(
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Panel/Services/ClassbookService.cs]: please remove space

File: CharlieBackend.Panel/Services/Interfaces/IClassbookService.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
+﻿using CharlieBackend.Core.DTO.Dashboard;
+using CharlieBackend.Panel.Models.Classbook;
+using System;
+using System.Threading.Tasks;
+
+namespace CharlieBackend.Panel.Services.Interfaces
+{
+    /// <summary>
+    /// Interface that describes methods for work with classbook and it's data.
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Panel/Services/Interfaces/IClassbookService.cs]: please fix it's => its

**Review by ArsenMaistryshyn — CHANGES_REQUESTED** [review]

File: CharlieBackend.Panel/Services/ClassbookService.cs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+        /// <summary>
+        /// Makes parallel asynchronous HTTP calls to API service and returns obtained data in a form
+        /// of a ViewModel that is ready to be viewed in a classbook.
+        /// </summary>
+        /// <param name="classbookFilter">DTO with an optional set of 
+        /// filters for classbook.</param>
+        /// <returns>>model with data required for viewing a classbook.</returns>
+        public async Task<ClassbookViewModel> GetClassbookAsync(
+            StudentsRequestDto<ClassbookResultType> classbookFilter)
```
Comment (BobTheSuperman) [inline CharlieBackend.Panel/Services/ClassbookService.cs]: add a second indent (+4 spaces)

File: CharlieBackend.Panel/Services/ClassbookService.cs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            var activeCourseDtos = await _apiUtil.GetAsync<IList<CourseDto>>(_getActiveCoursesEndpoint);
+
+            return _mapper.Map<IList<ClassbookCourseViewModel>>(activeCourseDtos);
+        }
+
+        private async Task<IList<ClassbookStudentsGroupViewModel>> GetStudentsGroupViewModelsAsync()
+        {
+            var studentGroupDtos = 
+                await _apiUtil.GetAsync<IList<StudentGroupDto>>(_getStudentGroupsEndpoint);
```
Comment (BobTheSuperman) [inline CharlieBackend.Panel/Services/ClassbookService.cs]: add a second indent (+4 spaces)

File: CharlieBackend.Panel/Services/ClassbookService.cs [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+            return _mapper.Map<IList<ClassbookStudentsGroupViewModel>>(studentGroupDtos);
+        }
+        
+        private async Task<FilteredClassbookViewModel> GetFilteredRegisterViewModelAsynk(
+            StudentsRequestDto<ClassbookResultType> groupsRegisterFilter)
+        {
+            var result = await _apiUtil.PostAsync< StudentsClassbookResultDto, StudentsRequestDto <ClassbookResultType>>(
+                url: _getStudentsClassbookEndpoint,
+                data: groupsRegisterFilter);
```
Comment (BobTheSuperman) [inline CharlieBackend.Panel/Services/ClassbookService.cs]: add a second indent (+4 spaces)

File: CharlieBackend.Panel/Services/ClassbookService.cs [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+                classbookFilter.CourseId = 1;
+            }
+            if (classbookFilter.StudentGroupId==null)
+            {
+                classbookFilter.StudentGroupId = 1;
+            }
+            if (classbookFilter.IncludeAnalytics==null)
+            {
+                classbookFilter.IncludeAnalytics = new ClassbookResultType[] {
+                ClassbookResultType.StudentPresence, ClassbookResultType.StudentMarks };
```
Comment (BobTheSuperman) [inline CharlieBackend.Panel/Services/ClassbookService.cs]: Please format.

**Review by BobTheSuperman — APPROVED** [review]

**Review by Ssashaa — APPROVED** [review]

File: CharlieBackend.Panel/Controllers/ClassbookController.cs (hunk 1 above)
Comment (DmitryTkachenko) [reply CharlieBackend.Panel/Controllers/ClassbookController.cs:3]: Done

File: CharlieBackend.Panel/Controllers/ClassbookController.cs (hunk 2 above)
Comment (DmitryTkachenko) [reply CharlieBackend.Panel/Controllers/ClassbookController.cs:4]: Done

File: CharlieBackend.Panel/Controllers/ClassbookController.cs (hunk 3 above)
Comment (DmitryTkachenko) [reply CharlieBackend.Panel/Controllers/ClassbookController.cs]: Done

File: CharlieBackend.Panel/Services/ClassbookService.cs (hunk 4 above)
Comment (DmitryTkachenko) [reply CharlieBackend.Panel/Services/ClassbookService.cs]: Done

File: CharlieBackend.Panel/Services/ClassbookService.cs (hunk 5 above)
Comment (DmitryTkachenko) [reply CharlieBackend.Panel/Services/ClassbookService.cs]: Done

File: CharlieBackend.Panel/Services/ClassbookService.cs (hunk 6 above)
Comment (DmitryTkachenko) [reply CharlieBackend.Panel/Services/ClassbookService.cs]: Done

File: CharlieBackend.Panel/Services/ClassbookService.cs (hunk 7 above)
Comment (DmitryTkachenko) [reply CharlieBackend.Panel/Services/ClassbookService.cs]: Done

File: CharlieBackend.Panel/Services/Interfaces/IClassbookService.cs (hunk 8 above)
Comment (DmitryTkachenko) [reply CharlieBackend.Panel/Services/Interfaces/IClassbookService.cs]: Done

File: CharlieBackend.Panel/Services/ClassbookService.cs (hunk 9 above)
Comment (DmitryTkachenko) [reply CharlieBackend.Panel/Services/ClassbookService.cs]: Done

File: CharlieBackend.Panel/Services/ClassbookService.cs (hunk 10 above)
Comment (DmitryTkachenko) [reply CharlieBackend.Panel/Services/ClassbookService.cs]: Done

File: CharlieBackend.Panel/Services/ClassbookService.cs (hunk 11 above)
Comment (DmitryTkachenko) [reply CharlieBackend.Panel/Services/ClassbookService.cs]: Done

File: CharlieBackend.Panel/Services/ClassbookService.cs (hunk 12 above)
Comment (DmitryTkachenko) [reply CharlieBackend.Panel/Services/ClassbookService.cs]: Done

**Review by da-real-ivanpan — APPROVED** [review]

**Review by Sergey-Kisenko — APPROVED** [review]

File: CharlieBackend.Panel/Models/Classbook/FilteredClassbookViewModel.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
+﻿using CharlieBackend.Core.DTO.Dashboard;
+using System;
+using System.Collections.Generic;
+
+namespace CharlieBackend.Panel.Models.Classbook
+{
+    public class FilteredClassbookViewModel
+    {       
+            public IList<StudentMarkDto> StudentsMarks { get; set; }
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Panel/Models/Classbook/FilteredClassbookViewModel.cs]: please remove one space

**Review by ArsenMaistryshyn — APPROVED** [review]

**Pull request merged**